chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,142 @@
+  - [ ] *Hdfs 3.x*: Introduced support for multiple active namenodes which allow for High availability and now, namenode recovery is done automatically.
+
+- [x] *Hdfs VS S3*: 
+  - [ ] *Hdfs*: A distributed file system that allows for storage of files in a hirarchiel FS, and allows for data querying and processing using tools such as MapReduce and Spark, in addition to the ability to utilize the GPU (as of hadoop 3.x) along with spark, however hdfs is usually self managed or managed using external vendors which come with a not so small cost, while the availability of these systems is lower than the availability of S3 which is managed by amazon and and is cheaper.
+        Hdfs also allows for querying on the data using apache Impala, Hive and Hbase 
+  - [ ] *S3* (Simple Storage Service): can only store and retrieved objects by their key / name, no querying can be done on the s3 and it is not a file system, meaning there are no folders, only file names, S3 allows to tag and store some additional metadata along with the file, has a very simple REST api that can be used and is managed by amazon, with other S3 compatible systems such as Minio that are also Object storage.
+        S3 is relatively cheap, as you only pay for what you use, in addition to it being managed by Amazon which have the resources and the knowledge to provide extremely high availability (11 nines) with good performance and a good cost.
+
+- [x] *Hdfs use cases*: to store large amounts of files that need to be searchable (Websites and text / pdf files), hadoop is often used in the finance industry for risk modelling (using big data that is stored in hdfs and analyzing with mapreduce and spark), can also be used like a regular file / block storage that can handle a lot of data and scale easily while being reliable, Hdfs can also be used along with analytics tools to detect fradulent activity.
+
+- [x] *Terms*:
+  - [x] *Failover*: The process of handling a node / master failure, when a leader fails, a failover event is triggered and a leader re election starts.
+  - [x] *Fencing*: Isolation of a node or shared resources when a node seems to be malfunctioning.
+  - [x] *Zkfc*:  The Zookeeper Failover Controller which is responsible for automatic failover in distributed systems like Hadoop (namenode failover) and Kafka (Controller failover)
+  - [x] *Quorum*: Quorum (in zookeeper) is the smallest amount of running servers of a distributed system in order for it to work and be available for the clients, usually the formula for a quorum is that if you have 2n + 1 nodes, n nodes can fail and the system will keep working as it worked before.
+  The quorum can also be used to elect leaders, by having a quorum of the in sync nodes that can become leaders in the case of node failure.
+
+
 1. **Q:** What is Apache Hadoop?
+1. **A:** Hadoop is a distributed file system that works based on a master-slave architecture which was built with the assumption that hardware failure is the norm.
```

Comment:
**Hadoop** is a distributed file system???
Check again most of this explanation is wrong and refers to the HDFS storage layer rather than hadoop as a whole.
Also please refer to computers as nodes/servers